shimmingtoolbox/optimizer/lsq_optimizer.py

```diff
@@ -225,15 +269,37 @@ def optimize(self, mask):
 
         # Define coil profiles
         n_channels = self.merged_coils.shape[3]
-
         mask_vec = mask.reshape((-1,))
-
+        
+        mask_erode = erode_binary_mask(mask,shape='sphere',size = 3)
+        mask_erode_vec = mask.reshape((-1,))
         # Reshape coil profile: X, Y, Z, N --> [mask.shape], N
         #   --> N, [mask.shape] --> N, mask.size --> mask.size, N --> masked points, N
-        coil_mat = np.reshape(np.transpose(self.merged_coils, axes=(3, 0, 1, 2)),
+        temp = np.transpose(self.merged_coils, axes=(3, 0, 1, 2))
+        coil_mat = np.reshape(temp,
                               (n_channels, -1)).T[mask_vec != 0, :]  # masked points x N
-        unshimmed_vec = np.reshape(self.unshimmed, (-1,))[mask_vec != 0]  # mV'
-
+        
+        merged_coils_Gx = np.zeros(np.shape(temp))
+        merged_coils_Gy = np.zeros(np.shape(temp))
+        merged_coils_Gz = np.zeros(np.shape(temp))
+        for ch in range(n_channels):
+            merged_coils_Gx[ch] = np.gradient(temp[ch], axis = 0)
+            merged_coils_Gy[ch] = np.gradient(temp[ch], axis = 1)
+            merged_coils_Gz[ch] = np.gradient(temp[ch], axis = 2)
+            
+        coil_Gx_mat = np.reshape(merged_coils_Gx,
+                              (n_channels, -1)).T[mask_erode_vec != 0, :]  # masked points x N
+        coil_Gy_mat = np.reshape(merged_coils_Gy,
+                              (n_channels, -1)).T[mask_erode_vec != 0, :]  # masked points x N
+        coil_Gz_mat = np.reshape(merged_coils_Gz,
+                              (n_channels, -1)).T[mask_erode_vec != 0, :]  # masked points x N
```

Comment:
What if the acquisition is oblique? The "z gradient" would be in the slice encoding direction, does that have an impact on your calculation? There is some code in the toolbox (shimmingtoolbox.shim.realtime_shim:realtime_shim()) to calculate the gradient in the "true z" direction.